### PR TITLE
- run 'stopContainer' in a Future (resolves #518)

### LINKED ISSUE #518: replace `WaitUtil.sleep` with Futures
right now if you specify a `wait` time, the plugin just basically sleeps which means if the container exits sooner you still have to wait for the sleep to complete.

what (hopefully) would be better would be to do is use a `Future` what would allow the process to continue as soon as it completes and timeout if it takes longer then the allotted `wait` time.

thoughts?

### PR DESCRIPTION
**Pull request by jgangemi**
resolves #518, now the plugin will continue as soon as the container stops, hurray faster builds!

### CODE REVIEW

**Comment by jgangemi** [comment]
@rhuss i'd appreciate consideration (and a release) at your earliest convenience. this shaved 10 minutes off our build.

**Comment by rhuss** [comment]
ok, will have a look tomorrow. The only possible problem I see here is, that I already have concurrency issues with Docker daemon running over unix socket (when waiting to a log output in a different thread), so I'm afraid that for this use case there might even more serious issues. However, for TCP based connection everything should be fine.

**Comment by jgangemi** [comment]
yeah - unix socket handling is crappy in general. i haven't looked at any of the libraries i saw to see if they were better then what we are using.

thinking about it now, it might not even need to run in a `Future` and we could just rely on the docker daemon to time itself out, but ultimately this is probably safer in that regard.

maybe we could add a `caveats` section that for the unix socket.

**Comment by rhuss** [comment]
oops, I mixed this up with #531 (where I see more serious concurrency issue coming). Guess the future is more or less sage ;-). to be continued tomorrow ...

File: src/main/java/io/fabric8/maven/docker/util/WaitUtil.java [hunk 1]
```diff
@@ -25,6 +25,8 @@
  */
 public class WaitUtil {
 
+    private static final int DEFAULT_DOCKER_TIMEOUT = 10;
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/util/WaitUtil.java]: 10s maybe a bit high as default, which is even used when no killGracePeriod is specified. In fact I wouldn't wait this no killGracePeriod is specified so simply skipping any wait when `int wait` is null.

Any particular reason why you always want a wait when a shutdown happens ?

File: src/main/java/io/fabric8/maven/docker/service/RunService.java [hunk 2]
```diff
@@ -412,32 +413,20 @@ private void shutdown(DockerAccess access,
...
         }
-        // Stop the container
-        int killGracePeriod = descriptor.getKillGracePeriod();
-        int killGracePeriodInSeconds = (killGracePeriod + 500) / 1000;
-        if (killGracePeriod != 0 && killGracePeriodInSeconds == 0) {
-            log.warn("A kill grace period of %d ms leads to no wait at all since its rounded to seconds. " +
-                     "Please use at least 500 as value for wait.kill",killGracePeriod);
-        }
-        access.stopContainer(containerId, killGracePeriodInSeconds);
-        if (killGracePeriod > 0) {
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/service/RunService.java]: There will be only a wait if killGracePeriod is > 0. I think we should do the same with the new logic, which means when no wait specified we shouldn't wait at all.

**Comment by rhuss** [comment]
LGTM except for the default value.

If don't mind I'll fix that right now and merge to integration. We can still discuss this till tomorrow. You can expect a release tomorrow or the day after.

File: src/main/java/io/fabric8/maven/docker/util/WaitUtil.java (hunk 1 above)
Comment (jgangemi) [inline src/main/java/io/fabric8/maven/docker/util/WaitUtil.java]: i don't always want to wait - the docker container takes up to 10 seconds which is why i used that default. if it exited before then, the future would complete, so i don't think we were waiting any extra time w/ this.

but whatever you think is best here.

Comment (jgangemi) [inline src/main/java/io/fabric8/maven/docker/util/WaitUtil.java]: yeah - looking at your change, i'm not 100% sure that's correct either in terms of returning `wait = 0` b/c you are always waiting some time for the daemon to complete, so i actually think the original implementation was correct.

```
get(long timeout, TimeUnit unit)
Waits if necessary for at most the given time for the computation to complete, and then retrieves its result, if available.```

if anything maybe it should be `if (wait == 0) { wait = DEFAULT }` (i can't remember what i originally had and i'm about to walk out the door for something).

**Pull request merged**